Add option to specify which content is examined for changes  (additions or deletions)

bwees opened the pull request:
This feature is described in #500

<img width="1584" alt="Screen Shot 2022-07-11 at 9 28 50 PM" src="https://user-images.githubusercontent.com/12686250/178388353-042d78d6-a518-4e33-9ef9-3d73fe3a05ab.png">

freddieleeman commented:
Not sure if this belongs here or has already been thought of, but I would like to see the `{diff}` only shows new additions, not changed or removed results. I think this makes a lot of sense in most situations and extends and complements the "trigger on additions"-feature.

Turning this:
```
(removed)     foo,
(removed)     foo2
(changed)     foo3
(into   )     bar
```
Into:
```
(added  )  bar
```

dgtlmoon commented:
> I think this makes a lot of sense in most situations and extends and complements the "trigger on additions"-feature.

You could go one step further and add a notification token `{diff_added}' `{diff_removed}` etc too, tho not important, just an idea

this is a great PR! and +100 points for the test case!

dgtlmoon commented on changedetectionio/fetch_site_status.py:
```diff
@@ -276,11 +277,28 @@ def run(self, uuid):
         # On the first run of a site, watch['previous_md5'] will be None, set it the current one.
         if not watch.get('previous_md5'):
             watch['previous_md5'] = fetched_md5
+            watch['previous_text'] = str(stripped_text_from_html)
```
I really dont like this, better would be to use the `.history` and examine the previous one if theres some special configuration that is enabled.. thoughts?

dgtlmoon reviewed (commented):
I dont like storing `previous_text` in the `watch` better would be to use the `.history` and examine the previous one if theres some special configuration that is enabled.. thoughts?

freddieleeman commented:
> {diff_added}

This is a great idea for complimenting the new "Only trigger when new lines appear"-feature. {diff_added} should have both `added` and `into` lines.

Nice thinking @dgtlmoon. FTW!

dgtlmoon commented:
@bwees quite a few test failures in other tests, so something low level is broken

dgtlmoon commented on changedetectionio/fetch_site_status.py:
```diff
@@ -276,11 +277,28 @@ def run(self, uuid):
         # On the first run of a site, watch['previous_md5'] will be None, set it the current one.
         if not watch.get('previous_md5'):
             watch['previous_md5'] = fetched_md5
+            watch['previous_text'] = str(stripped_text_from_html)
```
compare this with the previous in `.history` i think is better (if theres a setting enabled that requires it)

dgtlmoon replied:
@bwees hmm, but people will also setup like the CSS Filter, so they only want to know when certain parts of a website change (or have additions or removals)

I still think that this feature should not change this, it would be too confusing otherwise (related to filters)

but relating to triggers... hmmm

yes so one example would be, they only want to trigger
- when text is removed
- when text contains "xyz" 

but that should work by examining the last snapshot anyway? (without your custom snapshot saving)

bwees replied:
I am pretty sure that should not change because stripped_text_from_html is after the CSS extraction. As for the second part of your question, I am not sure how that would be affected. I think you would then have to make a hierarchy of tests that could be reordered in order to provide that functionality. (I feel that it is outside of the scope of this feature anyway) The case you provide would work if both of those conditions are true but not either/or.

bwees commented:
> @bwees quite a few test failures in other tests, so something low level is broken

I'm not sure what is broken... I followed a similar process for the feature #685. Looking at the output there are issues with basic_auth and other authentication test (which I haven't touched I don't think??). Are you able to look over my test file? It's my first time writing a test and I tried to follow a similar format to the one in #685. Here is the failing test output:

```

client = <FlaskClient <Flask 'changedetectionio'>>
live_server = <[AttributeError("'LiveServer' object has no attribute 'url'") raised in repr()] LiveServer object at 0x7f0893b2ad00>
    def test_basic_auth(client, live_server):
        live_server_setup(live_server)
        # Give the endpoint time to spin up
        time.sleep(1)
        # Add our URL to the import page
        test_url = url_for('test_basicauth_method', _external=True).replace("//","//myuser:mypass@")
        res = client.post(
            url_for("import_page"),
            data={"urls": test_url},
            follow_redirects=True
        )
        assert b"1 Imported" in res.data
        # Check form validation
        res = client.post(
            url_for("edit_page", uuid="first"),
            data={"css_filter": "", "url": test_url, "tag": "", "headers": "", 'fetch_backend': "html_requests"},
            follow_redirects=True
        )
        assert b"Updated watch." in res.data
        # Trigger a check
        client.get(url_for("form_watch_checknow"), follow_redirects=True)
        time.sleep(1)
        res = client.get(
            url_for("preview_page", uuid="first"),
            follow_redirects=True
        )
>       assert b'myuser mypass basic' in res.data
E       assert b'myuser mypass basic' in b'<!doctype html>\n<html lang="en">\n<head>\n    <meta charset="utf-8">\n    <meta name="viewport" content="width=devi...eed-icon.svg" height="15"></a>\n            </li>\n        </ul>\n    </div>\n</div>\n\n</section>\n\n</body>\n</html>'
E        +  where b'<!doctype html>\n<html lang="en">\n<head>\n    <meta charset="utf-8">\n    <meta name="viewport" content="width=devi...eed-icon.svg" height="15"></a>\n            </li>\n        </ul>\n    </div>\n</div>\n\n</section>\n\n</body>\n</html>' = <WrapperTestResponse 6969 bytes [200 OK]>.data
tests/test_auth.py:39: AssertionError
```

Looks to be an issue with the LiveServer on the test harness?

bwees commented:
I just ran the tests from your master branch on a Raspberry Pi. I am getting identical errors so I don't think its my code that is broken?

bwees commented:
Just rolled back to a [previous commit](https://github.com/dgtlmoon/changedetection.io/commit/46cb5cff66805c0eef184b56706995caaf1de7b8) of yours that had checks passing and I am still getting the same errors. Could it be an update to a library that broke this? I just installed all of the dev packages to the latest versions.

bwees commented:
Note: This is all on the master branch of dgtlmoon/changedetection.io 

So I fount this [issue for pytest-flask](https://github.com/pytest-dev/pytest-flask/issues/140) and I rolled my own version back to 1.1.0 as suggested by the thread. I am now getting errors such as: 

```
self = <Flask 'changedetectionio'>, rule = <Rule '/test-endpoint' (HEAD, OPTIONS, GET) -> test_endpoint>, endpoint = 'test_endpoint', view_func = <function live_server_setup.<locals>.test_endpoint at 0xb027d0c0>, provide_automatic_options = True
options = {'endpoint': 'test_endpoint'}, methods = {'GET', 'OPTIONS'}, required_methods = {'OPTIONS'}, old_func = <function live_server_setup.<locals>.test_endpoint at 0xb0c736a8>

    @setupmethod
    def add_url_rule(
        self,
        rule: str,
        endpoint: t.Optional[str] = None,
        view_func: t.Optional[t.Callable] = None,
        provide_automatic_options: t.Optional[bool] = None,
        **options: t.Any,
    ) -> None:
        if endpoint is None:
            endpoint = _endpoint_from_view_func(view_func)  # type: ignore
        options["endpoint"] = endpoint
        methods = options.pop("methods", None)
    
        # if the methods are not given and the view_func object knows its
        # methods we can use that instead.  If neither exists, we go with
        # a tuple of only ``GET`` as default.
        if methods is None:
            methods = getattr(view_func, "methods", None) or ("GET",)
        if isinstance(methods, str):
            raise TypeError(
                "Allowed methods must be a list of strings, for"
                ' example: @app.route(..., methods=["POST"])'
            )
        methods = {item.upper() for item in methods}
    
        # Methods that should always be added
        required_methods = set(getattr(view_func, "required_methods", ()))
    
        # starting with Flask 0.8 the view_func object can disable and
        # force-enable the automatic options handling.
        if provide_automatic_options is None:
            provide_automatic_options = getattr(
                view_func, "provide_automatic_options", None
            )
    
        if provide_automatic_options is None:
            if "OPTIONS" not in methods:
                provide_automatic_options = True
                required_methods.add("OPTIONS")
            else:
                provide_automatic_options = False
    
        # Add the required methods now.
        methods |= required_methods
    
        rule = self.url_rule_class(rule, methods=methods, **options)
        rule.provide_automatic_options = provide_automatic_options  # type: ignore
    
        self.url_map.add(rule)
        if view_func is not None:
            old_func = self.view_functions.get(endpoint)
            if old_func is not None and old_func != view_func:
                raise AssertionError(
>                   "View function mapping is overwriting an existing"
                    f" endpoint function: {endpoint}"
                )
E               AssertionError: View function mapping is overwriting an existing endpoint function: test_endpoint
```

bwees commented:
Ok the above error is because T was not using the test run script but after switching back to 1.1.0 the test errors go away. My test is failing so going to rewrite it now...

dgtlmoon commented:
Try these https://github.com/dgtlmoon/changedetection.io/blob/master/requirements-dev.txt this is what should work

Tests always pass in master for us here

bwees commented:
I have no idea what the issue was 🤷. Tests are running correctly now!

dgtlmoon commented:
> I have no idea what the issue was shrug. Tests are running correctly now!

err where?

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -178,3 +182,23 @@ def lines_contain_something_unique_compared_to_history(self, lines=[]):
+
+        # get diff types using difflib
+        cruncher = difflib.SequenceMatcher(isjunk=lambda x: x in " \\t", a=self.get("previous_text", ""), b=new_text)
+        for tag, alo, ahi, blo, bhi in cruncher.get_opcodes():
```
im getting a exception when it hits this line
```
object of type 'bool' has no len()
```

bwees replied:
Yeah I found the problem. The default value in the model was set to False. I changed to a empty string. Changes coming now

bwees commented:
I am still fixing up my test. Will be coming shortly. Should I plan to switch to self.history or keep the `previous_text`? I do not think this feature will work without `previous_text`. For example, if text on a page is deleted and then readded and you only have additions enabled, you will miss the trigger if the text is readded in the next snapshot.

dgtlmoon commented:
>  if text on a page is deleted and then readded and you only have additions enabled, you will miss the trigger if the text is readded in the next snapshot.


hmmm but yeah.. unsure, it feels redundant to have two snapshots of the last state tho, why cant we use the most recent history? i still dont quite get it :/

dgtlmoon commented:
`test_block_while_text_present` fails

dgtlmoon commented:
> You are only saving the most recent history (via save_history_text in Watch.py) if there is a trigger. This feature filters out the trigger and not the notification. Thus by filtering out any deletions, we do not save any snapshots that have deletions.

Alright I think i get it :D so yeah, dump it into an extra text file for now

bwees commented:
Ok perfect. Do you see any reason why `test_block_while_text_present` is failing?

dgtlmoon commented:
> Ok perfect. Do you see any reason why `test_block_while_text_present` is failing?

theres always a reason :)

```
FAILED tests/test_block_while_text_present.py::test_check_block_changedetection_text_NOT_present - assert b'unviewed' in b'<!doctype html>\n<html lang="en">\n<head>\n    <meta charset="ut...
```

looks like its showing up as CHANGED when it shouldnt be, or vise-versa

bwees commented:
Am I thinking of assertions wrong? I though that `assert b'unviewed' in ...` means that it will error out when unviewed is not found?

bwees commented:
Am I right in setting the default settings to True in Watch.py with the below code? 
```
'trigger_on_add': True, # Allow a trigger if there are additions from the last snapshot
'trigger_on_del': True, # Allow a trigger if there are deletions from the last snapshot
```
I just added some debug messages and when I pull the data values during a test, I am getting False for them. I believe this is the cause for the `test_block_while_text_present` failures.

bwees commented:
Does not seem like the default values in Watch.py are being used as default. Do I need to add a default value to the watch creation API to set these by default?

bwees commented:
Ok so I have isolated where the filters are being turned off for some reason during the `test_block_while_text_present.py` test. The lines that are updating the data store is here https://github.com/dgtlmoon/changedetection.io/blob/ebb8b8862115c8eb8f9ab5206463a80a414f8677/changedetectionio/__init__.py#L619

`form.data` is carrying this information: 
```... 'trigger_on_add': False, 'trigger_on_del': False, ...```

when I look at the supplied information to the creation of the `form` object
https://github.com/dgtlmoon/changedetection.io/blob/ebb8b8862115c8eb8f9ab5206463a80a414f8677/changedetectionio/__init__.py#L570
the following is given as parameters:
`formdata = {'text_should_not_be_present': 'out of stoCk\r\nfoobar', 'url': 'http://localhost:5005/test-endpoint', 'fetch_backend': 'html_requests'}`
and
`data = ... 'trigger_on_add': True, 'trigger_on_del': True, ...`

It looks like wtforms is not combining the form parts properly and outputting `'trigger_on_add': False, 'trigger_on_del': False`

I have implemented a fix in the commit above but I'm not sure if I should be touching this part. Am I missing something?

bwees commented:
Ok so I am getting frustrated with test system. I have all of my tests working on my computer (look at picture below for proof!, FYI the JSON failing to save was because of a file perm error on my NAS) but when it gets uploaded to github, the test that I wrote fails. I have verified that all tests complete without my tests in the suite. If someone more experienced with this project's test framework would like to make a test that actually works (I can't fix a failing test that works on my computer and then not another 🤷 ) that would probably be the best route. 

<img width="809" alt="Screen Shot 2022-07-12 at 9 29 30 PM" src="https://user-images.githubusercontent.com/12686250/178628980-9a124671-55e8-4eab-8fbf-b7aeea619576.png">

bwees commented:
> I have implemented a fix in the commit above but I'm not sure if I should be touching this part. Am I missing something?

I have found that there is a problem with this fix. I will be working on it today. I am going to change the UI around to accommodate the current tests. This will be changed instead from checkboxes that are always on to being a set of radio buttons such as:

**Do not trigger when change includes only this type of change:**
 - Do Not Filter Based on Change
 - Additions
 - Deletions

bwees commented:
@dgtlmoon I believe that I have fixed all of the issues with this feature, implemented tests that actually pass consistently, and haven't broken anything test wise (hopefully 😅). Sorry for all the back and forth. This is my first big feature that I have made a PR for and the first one I have had to work with a test suite!

dgtlmoon commented:
Almost perfect

One thing -


> Do not trigger when change includes only this type of change:
> - Do Not Filter Based on Change
> - [ ] Additions
> - [ ] Deletions

this logic is hard to parse in my brain and probably others, is it possible...


###  Only trigger change detection on
 - [ ] Additions
 - [ ] Deletions
 - [x] Anything

bwees commented:
This is the UI that I implemented (sorry forgot to update the post!)
<img width="1760" alt="Screen Shot 2022-07-14 at 10 10 13 AM" src="https://user-images.githubusercontent.com/12686250/179002497-995ca1d8-92fc-43d0-8759-6b9e0150fca2.png">

Where the options are (`Any Changes` being the default):
<img width="171" alt="Screen Shot 2022-07-14 at 10 10 47 AM" src="https://user-images.githubusercontent.com/12686250/179002629-aaaacae8-6554-4fc9-88b2-bd0b1fe10c28.png">

dgtlmoon commented on changedetectionio/templates/edit.html:
```diff
@@ -156,6 +156,15 @@
+                        <div class="pure-control-group" id="request-method">
+                            {{ render_field(form.trigger_type) }}
+                            <span class="pure-form-message-inline">
+                                This filter only compares to the previous snapshot. This is different from the "Only trigger when new lines appear" setting which checks all previous snapshots.
```
```suggestion
                                This filter only compares to the previous snapshot. This is different from the "Only trigger when new lines appear" setting which checks all previous snapshots.</br>
```

dgtlmoon commented on changedetectionio/templates/edit.html:
```diff
@@ -156,6 +156,15 @@
+                            {{ render_field(form.trigger_type) }}
+                            <span class="pure-form-message-inline">
+                                This filter only compares to the previous snapshot. This is different from the "Only trigger when new lines appear" setting which checks all previous snapshots.
+                                This is good for filtering out notifications that only are showing deleted content. (i.e A notice being removed from a website)
```
```suggestion
                                This is good for filtering out notifications that only are showing deleted content. (i.e A notice being removed from a website)</br>
```

dgtlmoon replied:
Stops the page from pushing to 100% width

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -151,6 +153,43 @@ def save_history_text(self, contents, timestamp):
+        output_path = "{}/{}".format(self.__datastore_path, self['uuid'])
+
+        snapshot_fname = "{}/previous.txt".format(output_path)
+        if not path.isfile(snapshot_fname):
```
Please don't do this kind of 'mystery logic' where it does something that's not written in the function name, it's best to fix it so that this situation never happens (like check `watch.history_n > 2`)

bwees replied:
Ok I will change it to check the history length first and if it does not exist just return an empty string

dgtlmoon replied:
better to let it throw an exception i think, that case should never happen

bwees replied:
If I throw an exception, it will always throw the exception since we are not saving the previous text until after it determines if a change was done or not. Going to keep it to returning an empty string but only `if self.history_n < 1` . It will no longer create the file if it does not exist

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -151,6 +153,43 @@ def save_history_text(self, contents, timestamp):
+
+        with open(snapshot_fname, 'wb') as f:
+            f.write(contents)
+            f.close()
```
not needed, `with` will close this
```suggestion

```

dgtlmoon commented:
I'm still not convinced that you need any of the extra `get_previous_text()` `save_previous_text()` stuff that you added, when I edit the code so..

```
--- a/changedetectionio/model/Watch.py
+++ b/changedetectionio/model/Watch.py
@@ -177,9 +177,7 @@ class model(dict):
     def get_previous_text(self):
         from os import path
 
-        output_path = "{}/{}".format(self.__datastore_path, self['uuid'])
-
-        snapshot_fname = "{}/previous.txt".format(output_path)
+        snapshot_fname = self.history[self.newest_history_key]
         if not path.isfile(snapshot_fname):
             self.save_previous_text(b'')
             return ""
```

all of your tests pass..

bwees commented:
> I'm still not convinced that you need any of the extra `get_previous_text()` `save_previous_text()` stuff that you added, when I edit the code so..
> 
> ```
> --- a/changedetectionio/model/Watch.py
> +++ b/changedetectionio/model/Watch.py
> @@ -177,9 +177,7 @@ class model(dict):
>      def get_previous_text(self):
>          from os import path
>  
> -        output_path = "{}/{}".format(self.__datastore_path, self['uuid'])
> -
> -        snapshot_fname = "{}/previous.txt".format(output_path)
> +        snapshot_fname = self.history[self.newest_history_key]
>          if not path.isfile(snapshot_fname):
>              self.save_previous_text(b'')
>              return ""
> ```
> 
> all of your tests pass..
> 
> I'de like to merge this in ASAP but i'm just not convinced that you need this extra text file.. send me some examples that would change my mind :)

One example that is the reason I made the text file in the first place is take my school closing example from #500. This example is if we use the history from the watch that is currently in master. If my school posts “school closed” and I have only addition triggers enabled and then they remove the notice (a deletion so it will not trigger and will not save to the watch’s history), when they post “school closed” again, there will be no change detected because when it goes to check what parts of the diff there are, there will be no change, deletions, or additions since we didn’t save to the history when it was originally deleted. I can test with the history but I don’t think it will work for the above reasoning

dgtlmoon commented:
>  This example is if we use the history from the watch that is currently in master. If my school posts “school closed” and I have only addition triggers enabled and then they remove the notice (a deletion so it will not trigger and will not save to the watch’s history), when they post “school closed” again, there will be no change detected because when it goes to check what parts of the diff there are, there will be no change, deletions, or additions since we didn’t save to the history when it was originally deleted. I can test with the history but I don’t think it will work for the above reasoning


Any chance you could add that to a test? :pray:

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -151,6 +153,42 @@ def save_history_text(self, contents, timestamp):
         #@todo bump static cache of the last timestamp so we dont need to examine the file to set a proper ''viewed'' status
         return snapshot_fname
 
+    # Save previous text snapshot for diffing
```
```suggestion
    # Save previous text snapshot for diffing - used for calculating additions and deletions
```

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -151,6 +153,42 @@ def save_history_text(self, contents, timestamp):
+
+        return snapshot_fname
+
+    # Get previous text snapshot for diffing
```
```suggestion
    # Get previous text snapshot for diffing - used for calculating additions and deletions
```

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -151,6 +153,42 @@ def save_history_text(self, contents, timestamp):
 
+    # Save previous text snapshot for diffing
+    def save_previous_text(self, contents):
+        from os import mkdir
```
```suggestion
      
```

not needed - we have `import os` at the top

dgtlmoon commented on changedetectionio/forms.py:
```diff
@@ -343,6 +343,7 @@ class watchForm(commonSettingsForm):
     check_unique_lines = BooleanField('Only trigger when new lines appear', default=False)
     trigger_text = StringListField('Trigger/wait for text', [validators.Optional(), ValidateListRegex()])
     text_should_not_be_present = StringListField('Block change-detection if text matches', [validators.Optional(), ValidateListRegex()])
+    trigger_type = SelectField(u'Trigger On', choices=[("all", "Any Changes"), ("add", "Only Additions"), ("delete", "Only Deletions")], coerce=str, default="add")
```
i'm fairly happy with SelectField, it's good enough for now :)

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -1,5 +1,6 @@
 import os
 import uuid as uuid_builder
+import difflib
```
Can go in `def get_diff_types()`

dgtlmoon reviewed (commented):
small changes

bwees commented:
> > This example is if we use the history from the watch that is currently in master. If my school posts “school closed” and I have only addition triggers enabled and then they remove the notice (a deletion so it will not trigger and will not save to the watch’s history), when they post “school closed” again, there will be no change detected because when it goes to check what parts of the diff there are, there will be no change, deletions, or additions since we didn’t save to the history when it was originally deleted. I can test with the history but I don’t think it will work for the above reasoning
> 
> Any chance you could add that to a test? 🙏

Yes I will add this to a new test. 

As for if we still need to include the previous_text.txt, we do. Here is the case that fails when using the history instead of previous_text.txt (The watch is set to only trigger on additions):

Content 1 (Start):
```
Foo1
Foo2
Foo3
```
Content 2 (Change IS Detected ✅ ):
```
Foo1
Foo2
Foo4
Foo3
```
Content 3 (Change IS NOT Detected ✅ ):
```
Foo1
Foo2
Foo3
```
Content 4 (Change is NOT Detected ❌ - Should trigger because we added `Foo4`.)
```
Foo2
Foo4
Foo3
```

Content 4 fails because it is comparing to the last time a change was detected (Content 2) which results in only deletions found which is filtered out. By using previous_text.txt, we are comparing to Content 3 which results in a correct trigger.

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -178,3 +213,25 @@ def lines_contain_something_unique_compared_to_history(self, lines=[]):
+        }
+
+        # get diff types using difflib
+        cruncher = difflib.SequenceMatcher(isjunk=lambda x: x in " \\t", a=str(self.get_previous_text()), b=str(new_text))
```
Does this also ignore line-feeds?

bwees replied:
what do you mean by line feeds?

dgtlmoon replied:
`isjunk=lambda x: x in " \\t"` so what if a page adds a line-feed, should that get included as a change?

bwees replied:
I am using the same algorithm that is used for notification generation. I think that only if text is added that a change should be considered?

dgtlmoon commented:
question, if 'diff filter' is set to 'any' (not Additions or Deletions) - does it still create the `previous text file` ? I guess it has to right? (in the case that you change the option)

bwees commented:
> question, if 'diff filter' is set to 'any' (not Additions or Deletions) - does it still create the `previous text file` ? I guess it has to right? (in the case that you change the option)

Yes this is correct. It creates the text file at the same time that it saves the previous md5 hash

dgtlmoon commented:
> > question, if 'diff filter' is set to 'any' (not Additions or Deletions) - does it still create the `previous text file` ? I guess it has to right? (in the case that you change the option)
> 
> Yes this is correct. It creates the text file at the same time that it saves the previous md5 hash

I'm getting close to being satisfied enough to merge this in, just thinking - for existing watches, what could be the impact? if the "previous text" file doesnt exist or is empty?

bwees commented:
> > > question, if 'diff filter' is set to 'any' (not Additions or Deletions) - does it still create the `previous text file` ? I guess it has to right? (in the case that you change the option)
> > 
> > 
> > Yes this is correct. It creates the text file at the same time that it saves the previous md5 hash
> 
> I'm getting close to being satisfied enough to merge this in, just thinking - for existing watches, what could be the impact? if the "previous text" file doesnt exist or is empty?

This shouldn't be affected, I will do a test and report back. I may have to add a line for the get text to say if the text file does not exist, also return an empty string. Is there a specific spot that I can add a migration step to?

dgtlmoon commented:
> This shouldn't be affected, I will do a test and report back. I may have to add a line for the get text to say if the text file does not exist, also return an empty string. Is there a specific spot that I can add a migration step to?

If you need it

https://github.com/dgtlmoon/changedetection.io/blob/a95468be089a8e94e83a2e83001e883f82001f2d/changedetectionio/store.py#L524

just add a new `update_n`

bwees commented:
> > This shouldn't be affected, I will do a test and report back. I may have to add a line for the get text to say if the text file does not exist, also return an empty string. Is there a specific spot that I can add a migration step to?
> 
> If you need it
> 
> https://github.com/dgtlmoon/changedetection.io/blob/a95468be089a8e94e83a2e83001e883f82001f2d/changedetectionio/store.py#L524
> 
> just add a new `update_n`

Perfect thanks

bwees commented:
I am going to add a migration step to prevent any weird errors or bugs that may result with returning an empty string if the text file is not found. I am going to set the contents of the file to the latest history snapshot.

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -151,6 +152,40 @@ def save_history_text(self, contents, timestamp):
+
+        with open(snapshot_fname, 'rb') as f:
+            contents = f.read()
+            f.close()
```
f.close not needed
```suggestion
```

dgtlmoon commented on changedetectionio/store.py:
```diff
@@ -527,3 +527,22 @@ def update_3(self):
+                with open(os.path.join(self.datastore_path, uuid, "previous.txt"), "w") as f:
+                    # Fill it with the latest history
+                    latest_file_name = watch.history[watch.newest_history_key]
+                    with open(latest_file_name, "r") as f2:
```
should be `rb` I believe (UTF-8 data)

dgtlmoon replied:
```suggestion
                    with open(latest_file_name, "rb") as f2:
```

dgtlmoon commented on changedetectionio/store.py:
```diff
@@ -527,3 +527,22 @@ def update_3(self):
+            # Check if the previous.txt exists
+            if not os.path.exists(os.path.join(self.datastore_path, uuid, "previous.txt")):
+                # Generate a previous.txt
+                with open(os.path.join(self.datastore_path, uuid, "previous.txt"), "w") as f:
```
should be `wb`
```suggestion
                with open(os.path.join(self.datastore_path, uuid, "previous.txt"), "wb") as f:
```

dgtlmoon reviewed (commented):
file IO changes

dgtlmoon commented:
https://github.com/dgtlmoon/changedetection.io/issues/500#issuecomment-1186956812

bwees commented:
I fixed merge conflicts with [7ac0620](https://github.com/dgtlmoon/changedetection.io/pull/749/commits/7ac062009917ebc5dc7409158ca9257b59cfcc31) so you can merge my version which keeps the change from the newest version.

bwees commented:
Not sure why tests aren't showing up but here is the run
https://github.com/bwees/changedetection.io/runs/7571166132?check_suite_focus=true

dgtlmoon reviewed (commented):
thanks!

dgtlmoon commented:
@bwees fantastic! could you fix that one last merge conflict then i'll take a look?

bwees commented:
> @bwees fantastic! could you fix that one last merge conflict then i'll take a look?

See https://github.com/dgtlmoon/changedetection.io/pull/749/commits/7ac062009917ebc5dc7409158ca9257b59cfcc31

dgtlmoon commented:
> > @bwees fantastic! could you fix that one last merge conflict then i'll take a look?
> 
> See [7ac0620](https://github.com/dgtlmoon/changedetection.io/commit/7ac062009917ebc5dc7409158ca9257b59cfcc31)

![image](https://user-images.githubusercontent.com/275001/181744021-ae1303c5-6d81-4c70-9693-2a5378ad0c79.png)


it still tells me theres a conflict

bwees commented:
Yes because any changes I do will result in a merge conflict. You must accept my changes as the correct version. I added back the line in my version that is resulting in the conflict. If you look at the diff between the proposed changes and the current version you will see that I added back the line that is present in the function above.

Brandon Wees
On Jul 29, 2022, 6:50 AM -0400, dgtlmoon ***@***.***>, wrote:
> > > @bwees fantastic! could you fix that one last merge conflict then i'll take a look?
> > See 7ac0620
> it still tells me theres a conflict
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>

dgtlmoon commented:
Hmm it lets me select 'nothing', I think it maybe needs a tiny bit of validation here (you can add validation at the form level in wtforms)

![image](https://user-images.githubusercontent.com/275001/181906380-5f9b335b-f573-47ac-97f8-09dea125e78b.png)

I'm wondering if this makes more sense (label change)

![image](https://user-images.githubusercontent.com/275001/181906442-cddc09aa-ace0-4499-81ad-e1333b4c4977.png)

because if i select `'deletions'` then I would expect any change detection to only show `deletions` "Limit filters and triggers to.." means that it doesnt affect the output, but only triggers right?

dgtlmoon commented:
Another thing, for example is that we need to be _very_ sure about additions and deletions, like here I first added a page to watch

https://keygenchurch.bandcamp.com/merch/keygen-church-t-shirt-datat11

Then I added a filter of `/html/body/div[5]/div/div[1]/div[2]/div[3]`

Then I see in my email

```
(changed) 
(changed)       Top
(changed)           Bandcamp
(changed)           * log in
(changed)           * terms of use
(changed)           * privacy
(changed)           * copyright policy
(changed)           * switch to mobile view
(changed)           * language: English
(changed)               + 日本語
(changed)               + Français
(into   )     Quality black unisex T-shirt 155g/m2 screen printed in Italy. Shipping from France. Available in MS-DOS Black.
(into   )     These T-shirts follow European sizing charts so for American customers in doubt, consider the size above what you are used to. IE L == XL.
(into   )     Size guide available here: [www.dataairlines.net/STTM528.pdf](http://www.dataairlines.net/STTM528.pdf)
(into   ) 
(into   )     Please note that all orders placed via bandcamp are shipped with unregistered shipping. If you need tracking or prefer shipping options that cover for loss or damage in transit, please visit our store at [www.dataairlines.net](http://www.dataairlines.net/)
(into   ) 
(into   )       Sold Out
(into   )
```

that it's telling me all of that was `changed` and not `removed`

so maybe the built in python diff algo isnt so good, none of that was 'changed' it was all removed

bwees commented:
> Another thing, for example is that we need to be _very_ sure about additions and deletions, like here I first added a page to watch
> 
> https://keygenchurch.bandcamp.com/merch/keygen-church-t-shirt-datat11
> 
> Then I added a filter of `/html/body/div[5]/div/div[1]/div[2]/div[3]`
> 
> Then I see in my email
> 
> ```
> (changed) 
> (changed)       Top
> (changed)           Bandcamp
> (changed)           * log in
> (changed)           * terms of use
> (changed)           * privacy
> (changed)           * copyright policy
> (changed)           * switch to mobile view
> (changed)           * language: English
> (changed)               + 日本語
> (changed)               + Français
> (into   )     Quality black unisex T-shirt 155g/m2 screen printed in Italy. Shipping from France. Available in MS-DOS Black.
> (into   )     These T-shirts follow European sizing charts so for American customers in doubt, consider the size above what you are used to. IE L == XL.
> (into   )     Size guide available here: [www.dataairlines.net/STTM528.pdf](http://www.dataairlines.net/STTM528.pdf)
> (into   ) 
> (into   )     Please note that all orders placed via bandcamp are shipped with unregistered shipping. If you need tracking or prefer shipping options that cover for loss or damage in transit, please visit our store at [www.dataairlines.net](http://www.dataairlines.net/)
> (into   ) 
> (into   )       Sold Out
> (into   )
> ```
> 
> that it's telling me all of that was `changed` and not `removed`
> 
> so maybe the built in python diff algo isnt so good, none of that was 'changed' it was all removed

So I found the case that produces a "changed" result. This occurs when the new text contains all new lines from the previous text. I will add an option and instead of labeling as 'Changes`, I will label as `All New Text`. Does this seem like an adequate solution or should I make it so if the lib detects a change we count it as an addition and deletion to preserve the current functionality?

| Original Text     | New Text                  | Result                             |
|-------------------|---------------------------|------------------------------------|
| Here<br>is<br>some<br>text | Here<br>is<br>some<br>changed<br>text | add: True, del: False, change: False |
| Here<br>is<br>some<br>text | All<br>new<br>text              | add: False, del: False, change: True |

dgtlmoon commented:
> I will label as `All New Text`. Does this seem like an adequate solution or should I make it so if the lib detects a change we count it as an addition and deletion to preserve the current functionality?

What about `All new and changed text` ?

bwees commented:
I made the form look like this in [0efc504](https://github.com/dgtlmoon/changedetection.io/pull/749/commits/0efc504c5dad8f81622c0ae2dd04e3d734fcffa5). I changed the wording slightly to be more verbose but let me know what you think.
<img width="460" alt="Screen Shot 2022-07-30 at 8 47 50 AM" src="https://user-images.githubusercontent.com/12686250/181914984-975442ad-f4d6-4376-b4fa-35fa6eb66e0d.png">

dgtlmoon commented:
> I made the form look like this in [0efc504](https://github.com/dgtlmoon/changedetection.io/pull/749/commits/0efc504c5dad8f81622c0ae2dd04e3d734fcffa5). I changed the wording slightly to be more verbose but let me know what you think. <img alt="Screen Shot 2022-07-30 at 8 47 50 AM" width="460" src="https://user-images.githubusercontent.com/12686250/181914984-975442ad-f4d6-4376-b4fa-35fa6eb66e0d.png">

hmm should be

[*] Additions and Changed text
[*] Deletions

If i'm correct that this will cover all cases?

bwees commented:
> > I will label as `All New Text`. Does this seem like an adequate solution or should I make it so if the lib detects a change we count it as an addition and deletion to preserve the current functionality?
> 
> What about `All new and changed text` ?

So rename `Additions` to this or make a new option? I feel like `changed text` in this situation is misleading since it will not be a "change" unless there is all new text. When I think of changed text I think of a single line having something added or deleted from it which is not what "changed" means in this situation.

dgtlmoon commented:
I dont know how to reply to your question then, I would just refer back to https://github.com/dgtlmoon/changedetection.io/pull/749#issuecomment-1200144509

You need to be sure that people never ever lose data because of a misunderstanding in the option and what the "diff" algorithm does

bwees commented:
I feel that if we keep the current "Additions" and "Deletions" that we have now and just make the algorithm count a change as both addition and deletion this would standardize the algorithm.

dgtlmoon commented:
> So rename Additions to this or make a new option? I feel like changed text in this situation is misleading since it will not be a "change" unless there is all new text. 

in my original comment i pointed out that this is not true, because it's counting removals as changes... so i'm too confused about the whole thing now, i feel like python diff cant be trusted to line up with those checkbox options

dgtlmoon commented:
I dont know how to continue in this PR, i'm totally confused.

bwees commented:
If you look at the table that i have below, you will see that in the first example, it only counts additions and deletions. In the second row I have text that is completely different (new words) from the first one. You can see that it counts this as only a change and not deletions and additions. This is because we have completely new text which is what you experienced when you added a filter of `/html/body/div[5]/div/div[1]/div[2]/div[3]` which produced completely new text from when you originally were capturing the entire page (no HTML filter). This is why the algorithm was counting it as a change and not a deletion and addition. If I change it to be so a "change" from the diff algorithm produces  additions and deletions instead of just a change, the program will work as you expected in https://github.com/dgtlmoon/changedetection.io/pull/749#issuecomment-1200144509. 

I have not touched the preview generation that gets sent in the notification so that will continue to work how you experienced. However, you did bring to light a bug that would have affected this feature where if all of the text was different, a notification would never have been sent.

bwees commented:
Try using the feature now with this commit [6bdcdfb](https://github.com/dgtlmoon/changedetection.io/pull/749/commits/6bdcdfbaea7b817a09a492556164d3eea2313313) and ignore what the diff preview shows as a change in the notification. This feature will now count a "replace" (which produces the `(changed)` and `(into) `tags for the preview) as a addition and deletion which will be identical to how the diff algorithm works when it does not see completely new text.

dgtlmoon commented:
> However, you did bring to light a bug that would have affected this feature where if all of the text was different and you had one of the checkboxes deselected, a notification would never have been sent.

please add a test for this case if not already

bwees commented:
@dgtlmoon Do you see any other things that need fixed in this PR?

dgtlmoon commented:
@bwees still getting there, quite a lot of other fires to put out atm :/ just fixed some more conflicts that came up recently

bwees commented:
@dgtlmoon looks like a couple of import statements were removed during the merge. Looks like everything got fixed in the tests.

dgtlmoon commented on changedetectionio/tests/test_diff_filter_only_additions.py:
```diff
@@ -0,0 +1,83 @@
+
+def set_original_response():
+    test_return_data = """
+        A few new lines
```
I was thinking that more test data here could be good, like a much more test case perhaps

bwees replied:
are you saying adding a bunch to the single test case or adding a bunch more situations?

bwees commented:
Thanks! As for the comment in the review. Would you like more tests or just more content in the test?

bwees commented:
I am pretty sure that should not change because stripped_text_from_html is after the CSS extraction. As for the second part of your question, I am not sure how that would be affected. I think you would then have to make a hierarchy of tests that could be reordered in order to provide that functionality. (I feel that it is outside of the scope of this feature anyway) The case you provide would work if both of those conditions are true but not either/or.

Brandon Wees
On Jul 15, 2022, 2:53 PM -0400, dgtlmoon ***@***.***>, wrote:
> @dgtlmoon commented on this pull request.
> In changedetectionio/fetch_site_status.py:
> >          # Always record the new checksum
>         update_obj["previous_md5"] = fetched_md5
>
>         # On the first run of a site, watch['previous_md5'] will be None, set it the current one.
>         if not watch.get('previous_md5'):
>             watch['previous_md5'] = fetched_md5
> +            watch['previous_text'] = str(stripped_text_from_html)
> @bwees hmm, but people will also setup like the CSS Filter, so they only want to know when certain parts of a website change (or have additions or removals)
> I still think that this feature should not change this, it would be too confusing otherwise (related to filters)
> but relating to triggers... hmmm
> yes so one example would be, they only want to trigger
>
> • when text is removed
> • when text contains "xyz"
>
> but that should work by examining the last snapshot anyway? (without your custom snapshot saving)
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>

dgtlmoon commented:
@bwees i think the way forwards is to merge this into `master` but put a `beta` or something tag on the UI next to the field, the default behaviour doesnt change anything right?

dgtlmoon commented on changedetectionio/store.py:
```diff
@@ -578,8 +582,28 @@ def update_5(self):
     # We incorrectly used common header overrides that should only apply to Requests
     # These are now handled in content_fetcher::html_requests and shouldnt be passed to Playwright/Selenium
-    def update_7(self):
+    def update_6(self):
```
oops sorry, I made a bad conflict resolve, could you change this to

```suggestion
    def update_8(self):
```

bwees commented:
> @bwees i think the way forwards is to merge this into `master` but put a `beta` or something tag on the UI next to the field, the default behaviour doesnt change anything right?

Correct this does not change the default behavior. If there are issues, would you mind tagging me in the issue and I will do my best to fix?

dgtlmoon commented:
I might just quickly test that it doesnt accidently trigger some notifications when the site "upgrades", then it should be good to go

dgtlmoon commented on changedetectionio/store.py:
```diff
@@ -578,8 +582,28 @@ def update_5(self):
     # We incorrectly used common header overrides that should only apply to Requests
     # These are now handled in content_fetcher::html_requests and shouldnt be passed to Playwright/Selenium
-    def update_7(self):
+    def update_8(self):
```
sorry I meant this one, should be

```suggestion
    def update_7(self):
```

dgtlmoon commented on changedetectionio/store.py:
```diff
@@ -578,8 +582,28 @@ def update_5(self):
                 del self.data['settings']['headers'][v]
+                
+    # Generate a previous.txt for all watches that do not have one and contain history
+    def update_7(self):
```
sorry I meant this one, should be

```suggestion
    def update_8(self):
```

dgtlmoon commented:
> I feel that if we keep the current "Additions" and "Deletions" that we have now and just make the algorithm count a change as both addition and deletion this would standardize the algorithm.

This is not obvious/clear to the user


![image](https://user-images.githubusercontent.com/275001/196152161-76dc2bbe-3025-4d40-a506-228d4a468ffe.png)


What would happen with "changes" here?

bwees commented:
They would be counted no matter what check box was selected. This is the same behavior that is observed in the diffing algorithm

Brandon Wees
On Oct 17, 2022, 6:14 AM -0400, dgtlmoon ***@***.***>, wrote:
> > I feel that if we keep the current "Additions" and "Deletions" that we have now and just make the algorithm count a change as both addition and deletion this would standardize the algorithm.
> This is not obvious/clear to the user
> What would happen with "changes" here?
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>

dgtlmoon commented on changedetectionio/templates/edit.html:
```diff
@@ -172,6 +172,16 @@
+                            {{ render_checkbox_field(form.trigger_add, class="trigger-type") }}
+                            {{ render_checkbox_field(form.trigger_del, class="trigger-type") }}
+                            <span class="pure-form-message-inline">
+                                Filters the change-detection of this watch to only this content change.
```
```suggestion
                                Filters the change-detection of this watch to only this type of content change. <strong>replacements</strong> (neither additions or deletions) are always included.
```

dgtlmoon replied:
What do you think?

bwees replied:
Capitalize the “r” in replacements and then it looks good to me

bwees replied:
Resoved via [c80f463](https://github.com/dgtlmoon/changedetection.io/pull/749/commits/c80f46308aa87168142249c413e915963eb70bda)

dgtlmoon replied:
Cool thanks! I think it makes way more sense now, trying to think of situations that might confuse people (end up in more issues in the queue :) )

bwees replied:
Totally get it!

dgtlmoon commented:
@bwees heya, sorry, I added to the 'backups' test, can you change the `2` to a `3` and bum the note in the line above

https://github.com/bwees/changedetection.io/blob/99f3b01013904b9bc3a818e3cd21e17826500ba8/changedetectionio/tests/test_backup.py#L50

dgtlmoon commented:
@bwees please remove the headers request stuff from this branch

dgtlmoon commented:
@bwees do you see https://github.com/dgtlmoon/changedetection.io/pull/749/files/c80f46308aa87168142249c413e915963eb70bda#diff-2170be5b7a2abab3761e86bddf67b8f8fb71efbdc5a4fea3298100c9c0f1a360R181 ?

bwees commented:
> @bwees do you see https://github.com/dgtlmoon/changedetection.io/pull/749/files/c80f46308aa87168142249c413e915963eb70bda#diff-2170be5b7a2abab3761e86bddf67b8f8fb71efbdc5a4fea3298100c9c0f1a360R181 ?

Yes, the conversation is marked as resolved for me and I committed the changes in [c80f463](https://github.com/dgtlmoon/changedetection.io/pull/749/commits/c80f46308aa87168142249c413e915963eb70bda)

dgtlmoon commented:
> Sorry! Shows how much of a git noob I am. Sorry I will fix this.

Thanks! And just the one text change, do you see it at https://github.com/dgtlmoon/changedetection.io/pull/749/files/c80f46308aa87168142249c413e915963eb70bda#diff-2170be5b7a2abab3761e86bddf67b8f8fb71efbdc5a4fea3298100c9c0f1a360R181 ?

bwees commented on changedetectionio/templates/edit.html:
```diff
@@ -172,6 +172,16 @@
+                            {{ render_checkbox_field(form.trigger_add, class="trigger-type") }}
+                            {{ render_checkbox_field(form.trigger_del, class="trigger-type") }}
+                            <span class="pure-form-message-inline">
+                                Filters the change-detection of this watch to only this type of content change. <strong>Replacements</strong> (neither additions or deletions) are always included.
```
yeah this one isnt coming up for me. This is correct, will commit.

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -196,6 +198,39 @@ def save_history_text(self, contents, timestamp):
+        output_path = os.path.join(self.__datastore_path, self['uuid'])
+
+        # Incase the operator deleted it, check and create.
+        if not os.path.isdir(output_path):
```
You can remove this and call https://github.com/dgtlmoon/changedetection.io/blob/57f604dff1580ec49e58469edf616b83f0af9758/changedetectionio/model/Watch.py#L86

bwees replied:
Completed in https://github.com/dgtlmoon/changedetection.io/pull/749/commits/c2c8bb4de86af5209f36fee9e5b0902b74556bc9

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -196,6 +198,39 @@ def save_history_text(self, contents, timestamp):
+        if not os.path.isdir(output_path):
+            os.mkdir(output_path)
+
+        snapshot_fname = os.path.join(output_path, "previous.txt")
```
```suggestion
        snapshot_fname = os.path.join(self.watch_data_dir, "previous.txt")
```

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -196,6 +198,39 @@ def save_history_text(self, contents, timestamp):
+    def get_previous_text(self):
+        from os import path
+
+        output_path = os.path.join(self.__datastore_path, self['uuid'])
```
```suggestion

```

dgtlmoon commented on changedetectionio/model/Watch.py:
```diff
@@ -196,6 +198,39 @@ def save_history_text(self, contents, timestamp):
+
+        output_path = os.path.join(self.__datastore_path, self['uuid'])
+
+        snapshot_fname = os.path.join(output_path, "previous.txt")
```
```suggestion
        snapshot_fname = os.path.join(self.watch_data_dir, "previous.txt")
```

dgtlmoon commented on changedetectionio/store.py:
```diff
@@ -583,3 +587,23 @@ def update_7(self):
+
+
+            # Check if the previous.txt exists
+            if not os.path.exists(os.path.join(self.datastore_path, uuid, "previous.txt")):
```
```suggestion
            if not os.path.exists(os.path.join(watch.watch_data_dir, "previous.txt")):
```

dgtlmoon commented on changedetectionio/store.py:
```diff
@@ -583,3 +587,23 @@ def update_7(self):
+            # Check if the previous.txt exists
+            if not os.path.exists(os.path.join(self.datastore_path, uuid, "previous.txt")):
+                # Generate a previous.txt
+                with open(os.path.join(self.datastore_path, uuid, "previous.txt"), "wb") as f:
```
```suggestion
                with open(os.path.join(watch.watch_data_dir, "previous.txt"), "wb") as f:
```

dgtlmoon commented on changedetectionio/tests/test_diff_filter_only_deletions.py:
```diff
@@ -0,0 +1,73 @@
+
+    time.sleep(sleep_time_for_fetch_thread)
+    # Trigger a check
+    client.get(url_for("form_watch_checknow"), follow_redirects=True)
```
not needed because the `post` will trigger a recheck
```suggestion

```

dgtlmoon commented on changedetectionio/tests/test_diff_filter_changes_as_add_delete.py:
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/python3
+
```
```suggestion
+ @NOTE:  THIS RELIES ON SOME MIDDLEWARE TO MAKE CHECKBOXES WORK WITH WTFORMS UNDER TEST CONDITION, see changedetectionio/tests/util.py
```

good to add a little pointer which may save someone a lot of time in the future

dgtlmoon commented:
> I am still fixing up my test. Will be coming shortly. Should I plan to switch to self.history or keep the `previous_text`? I do not think this feature will work without `previous_text`. For example, if text on a page is deleted and then readded and you only have additions enabled, you will miss the trigger if the text is readded in the next snapshot.

aaah hmmmm

bwees commented:
Still says that changes are requested from me but when I click show changes it takes me nowhere, if there are any outstanding, let me know.

dgtlmoon commented on changedetectionio/store.py:
```diff
@@ -583,3 +587,23 @@ def update_7(self):
+            # Make sure we actually have history
+            if (watch.history_n == 0):
+                continue
+            latest_file_name = watch.history[watch.newest_history_key]
```
```suggestion
            
```

Doesnt seem todo anything, can be removed

dgtlmoon commented on changedetectionio/store.py:
```diff
@@ -583,3 +587,23 @@ def update_7(self):
+
+
+            # Check if the previous.txt exists
+            if not os.path.exists(os.path.join(watch.watch_data_dir, "previous.txt")):
```
```suggestion
            if os.path.exists(watch.watch_data_dir) and not os.path.exists(os.path.join(watch.watch_data_dir, "previous.txt")):
```

just to be sure

dgtlmoon commented on changedetectionio/tests/test_diff_filter_changes_as_add_delete.py:
```diff
@@ -0,0 +1,72 @@
+
+    # We should see the change
+    res = client.get(url_for("index"))
+    assert b'unviewed' in res.data
```
```suggestion
    assert b'unviewed' in res.data
    
    
    # Now check 'changes' are always going to be triggered
    set_original_response()
    client.post(
        url_for("edit_page", uuid="first"),
        # Neither trigger add nor del? then we should see changes still
        data={"trigger_add": "n",
              "trigger_del": "n",
              "url": test_url,
              "fetch_backend": "html_requests"},
        follow_redirects=True
    )
    time.sleep(sleep_time_for_fetch_thread)
    client.get(url_for("mark_all_viewed"), follow_redirects=True)
    
    # Same as original response but 'is->ix'
    test_return_data = """
        Here
        ix
        some
        text
    """

    with open("test-datastore/endpoint-content.txt", "w") as f:
        f.write(test_return_data)
        
    client.get(url_for("form_watch_checknow"), follow_redirects=True)
    time.sleep(sleep_time_for_fetch_thread)
    res = client.get(url_for("index"))
    assert b'unviewed' in res.data
```

Could you add this, that explicitly tests that changes (neither add or del) is triggered?

bwees commented:
> Could you add this, that explicitly tests that changes (neither add or del) is triggered?

So this test will always fail, when you disable trigger_add and trigger_del, no notifications will get through. This is an impossible case because there is form validation in the UI as seen below (when a form is submitted and both checkboxes are unchecked: 
<img width="402" alt="Screen Shot 2022-10-28 at 3 26 15 PM" src="https://user-images.githubusercontent.com/12686250/198716603-1218360b-bb62-4def-a8d6-0f741c0e0119.png">

I am going to roll back this test to the way it was before.

dgtlmoon commented:
> because there is form validation in the UI as seen below (when a form is submitted and both checkboxes are unchecked:

aaaah ok! yes thats right, that's why we were talking about radio button vs checkbox

bwees commented:
Unrelate but looks like its not only us dealing with the review changes bug https://github.com/community/community/discussions/9709

dgtlmoon commented:
@bwees thoughts on how this solves #1071 ? I think this pretty much nails it right?

dgtlmoon commented:
Make me wonder why/if we were doing it right before at all (by not checking the output of diff)

0xftoor commented:
@bwees & @dgtlmoon Looking forward for this to be pushed to master! Great work guys!

bwees commented:
> Make me wonder why/if we were doing it right before at all (by not checking the output of diff)

Not really sure what you mean by this but feel free to correct me. Are you asking why we don't just program it to apply the filters to the diff only instead of the whole page and put this setting in a radio button? As for that it still does not solve a problem like this:

I am running my own branch of changedetection with these features added (that's what happened with the other PR being merged with this one. oops!). I am doing this in part to battle-test the stuff I have written. I have a watch where it checks the output of a Recent updates page on a site. This is kind of like a queue that shows items that were updated recently. After 7 days of inactivity, the item disappears. I would like to turn off the notifications for when it is disappearing but still want to be notified when the item reappears.

> @bwees thoughts on how this solves #1071 ? I think this pretty much nails it right?

I don't think it does. I believe @peppetemp is trying to get notified when text is added that is not `No incidents reported`. This PR would solve the deletion of Oct. 29 in the example but not fix the notification of when it is added because they still want to be notified (just not when it says `No incidents reported`). THis would have to be a separate PR to achieve the functionality they want.

dgtlmoon commented:
thanks for the answers!
>  This is kind of like a queue that shows items that were updated recently. After 7 days of inactivity, the item disappears. I would like to turn off the notifications for when it is disappearing but still want to be notified when the item reappears.


So you're waiting for the item to be re-added and hoping to get the notification right?

bwees commented:
Yeah, so it's a watch of the recent updates table in the grade book my school uses. When a grade for one of my classes gets updated, the item gets moved to the top of the list. If the grade has not been updated for 7 days it gets removed from the list. By using the feature in this PR, I will not receive a notification when the item is removed (which is what I want) but I will get a notification when it is updated again (gets moved to the top) or when it is updated after not being updated for 7 days (reappears on the table)

dgtlmoon commented on changedetectionio/fetch_site_status.py:
```diff
@@ -293,8 +293,23 @@ def run(self, uuid):
-        # Always record the new checksum
+        if changed_detected:
+            if not watch.get("trigger_add", True) or not watch.get("trigger_del", True): # if we are supposed to filter any diff types
+                # get the diff types present in the watch
```
```suggestion
                # Get the diff types present in the watch
```

dgtlmoon commented on changedetectionio/fetch_site_status.py:
```diff
@@ -293,8 +293,23 @@ def run(self, uuid):
                     logging.debug("check_unique_lines: UUID {} had unique content".format(uuid))
 
-        # Always record the new checksum
+        if changed_detected:
```
```suggestion

```

can be removed (already under `if change_detected:`)

bwees commented:
> So "Add option to specify which content is examined for changes (additions or deletions)" is wrong
> 
> because here we are not examining those changes
> 
> here in this PR we are simply asking _"So a change was detected, was there a "add" or "delete" ?"_
> 
> right?
> 
> https://github.com/dgtlmoon/changedetection.io/pull/749/files#diff-13180567d89427347714f892e7d9b21724e3910632e867e0400fa8ee579ce9e8R302

Correct!

dgtlmoon commented:
> Correct!

hmmmm sorry, but this is really confusing functionality then, the PR title never matched what it was going todo exactly, this is going to lead to a lot of support issues I believe

I'm going to close this PR in favour of going in this direction https://github.com/dgtlmoon/changedetection.io/issues/1071

dgtlmoon commented:
btw, thanks for your hard work here, a lot of the code will get used to solve #1071

bwees commented:
Ok, that makes sense. Seems like a better way to function honestly. Good luck and no problem at all!

dgtlmoon commented:
Partially solved by https://github.com/dgtlmoon/changedetection.io/pull/1454 but this PR needs revisiting again

reecespieces commented:
> Not sure if this belongs here or has already been thought of, but I would like to see the `{diff}` only shows new additions, not changed or removed results. I think this makes a lot of sense in most situations and extends and complements the "trigger on additions"-feature.
> 
> Turning this:
> 
> ```
> (removed)     foo
> (removed)     foo2
> (changed)     foo3
> (into   )     bar
> ```
> 
> Into:
> 
> ```
> (added  )  bar
> ```

This is a lot more difficult than it initially seems.


Sometimes the (into) line will show new additions, but it will also trigger when content gets moved around.


So if you changed the (into) line into (added), it would show new content but it may also show content that was already on the page that just got moved somewhere else.
